dev/WindowsAppRuntime_UniversalBGTaskDLL/Task.cpp

```
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "Task.h"
+#if __has_include("Task.g.cpp")
+#include "Task.g.cpp"
+#endif
+
+using namespace winrt;
+using namespace winrt::Windows::ApplicationModel::Background;
+using namespace winrt::Windows::Storage;
+
+namespace winrt::Microsoft::Windows::ApplicationModel::Background::UniversalBGTask::implementation
+{
+    inline constexpr CLSID IID_IBackgroundTask = { 2098451764, 64786, 17358, 140, 34, 234, 31, 241, 60, 6, 223 };
+    void Task::Run(winrt::Windows::ApplicationModel::Background::IBackgroundTaskInstance taskInstance)
+    {
+        winrt::hstring lookupStr = winrt::to_hstring(taskInstance.Task().TaskId());
+        ApplicationDataContainer localSettings = ApplicationData::Current().LocalSettings();
+        auto values = localSettings.Values();
+        auto lookupobj = values.Lookup(lookupStr);
+        winrt::guid comClsId = winrt::unbox_value<winrt::guid>(lookupobj);
+
+        winrt::Windows::ApplicationModel::Background::IBackgroundTask bgTask = nullptr;
+        winrt::hresult hr = CoCreateInstance(comClsId, nullptr, CLSCTX_LOCAL_SERVER, IID_IBackgroundTask, reinterpret_cast<void**>(&bgTask));
+        bgTask.Run(taskInstance);
```

This is gonna AV if CoCreateInstance failed. and bgTask is still nullptr.

---

Thanks @jlaanstra for the review, This method gets called from Broker Infrastructure and not handle any error results, Isn't it fine to throw exception here in that case? It will kill the backgroundtaskhost.exe process that gets created by broker infrastructure in that case.

---

This won't throw an exception. You'll run into a nullref which will terminate. If you want to throw you should throw the hr.

---

Yes, It may not make much difference from BI side as the BI will just get to know that the backgroundtaskhost process got disconnected as it runs down. However I see that it is better to throw the right exception which may help during debugging.